File: benchmarks/benchmarks.cc
#include <algorithm>
#include <chrono>
#include <cstdio>
#include <mt19937.h>

// Neither GCC nor Clang eliminate the function call or loops while optimising.
// The definitions of the functions are in a shared object, and not visible to
// the compiler.
#define benchmark(function, iterations)  \
{  \
    auto delay = std::chrono::nanoseconds::max();  \
    for(int i = 0; i < 32; ++i)  \
    {  \
        auto begin = std::chrono::high_resolution_clock::now();  \
        for(int long i = 0; i < iterations; ++i)  \
        {  \
            function();  \
        }  \
        auto end = std::chrono::high_resolution_clock::now();  \
        auto delay_ = std::chrono::duration_cast<std::chrono::nanoseconds>(end - begin);  \
        delay = std::min(delay, delay_);  \
    }  \
    auto result = delay.count() / static_cast<double>(iterations);  \
    std::printf("%20s %8.2lf ns\n", #function, result);  \
}

/******************************************************************************
 * Main function.
 *****************************************************************************/
int main(void)
{
    benchmark(mt19937::init32, 0x1000L)
    benchmark(mt19937::init64, 0x1000L)
    benchmark(mt19937::rand32, 0xFFF0L)
    benchmark(mt19937::rand64, 0xFFF0L)
    benchmark(mt19937::real32, 0xFFF0L)
    benchmark(mt19937::real64, 0xFFF0L)
}

File: benchmarks/benchmarks.py
#! /usr/bin/env python3

import math
import mt19937
import time
import timeit


def benchmark(stmt, number, passes=32):
    delay = math.inf
    for _ in range(passes):
        delay_ = timeit.timeit(stmt=stmt, number=number, timer=time.perf_counter_ns)
        delay = min(delay, delay_)
    result = delay / number
    print(f'{stmt.__name__:>20} {result:8.2f} ns')


def main():
    """Main function."""
    benchmark(mt19937.init32, 0x1000)
    benchmark(mt19937.init64, 0x1000)
    benchmark(mt19937.rand32, 0xFFF0)
    benchmark(mt19937.rand64, 0xFFF0)
    benchmark(mt19937.real32, 0xFFF0)
    benchmark(mt19937.real64, 0xFFF0)


if __name__ == '__main__':
    main()
